File: matlab/raicesmult.m
format long

f=input('Ingrese la función f(x), por ejemplo: @(x) exp(-x)-x  :');
x0=input('Ingrese el valor inicial x0:  ');
n=input ('Ingrese el número de iteraciones máximo:  ');
tol=input ('Ingrese la tolerancia:  ');

[x,e,i] = mroots(f,x0,n,tol);
function [x0,er,i] = mroots(f,x0,n,tol)

syms x;
inicial=x0;
sp = '%.10f';
spa='%.10f %.10f %.10f %.10f %.10f %.10f\n';
disp (['i ' 'x ' 'fx ' ' f´(x)' ' f´´(x)' ' error abs ']);
fx=double(subs(f,x0));
df=matlabFunction( diff(f(x)));
d2f=matlabFunction( diff(df(x)));
dfx=double(subs(df,x0));
d2fx=double(subs(d2f,x0));
i=0;
den=dfx^2-(fx*d2fx);
er=tol+1;
while den~=0 && fx~=0 && er>tol && i<n
    if i==0
        er=nan;
    end
    resp=[i double(x0) double(fx) double(dfx) double(d2fx) double(er)];
    fprintf( spa, resp.' ) ;
    x1=x0-((fx*dfx)/den);
    fx=double(subs(f,x1));
    dfx=double(subs(df,x1));
    d2fx=double(subs(d2f,x1));
    den=dfx^2-(fx*d2fx);
    er=abs(x1-x0);
    x0=x1;
    i=i+1;
end
resp=[i double(x0) double(fx) double(dfx) double(d2fx) double(er)];
fprintf( spa, resp.' ) ;
if fx==0
    disp([num2str(x0,sp) ' es raíz.']);
elseif er<tol
    disp([num2str(x1,sp) ' es aproximación a una raíz con tolerancia ' num2str(tol,sp)]);
elseif den==0
    disp('No se pudo desarrollar el método puesto que al calcular el xn, el denominador es 0');
else
    disp(['El método fracasó con ' num2str(i) ' iteraciones.']);
end
ejex = linspace(inicial, x0+100*tol);
figure
plot(ejex, double(subs(f,ejex)))
grid
end

File: python/raicesmult.py
import sympy as sp

#INPUT


def rmult(x,f,x0,nIterations,tolerance):
    # METHOD SETUP

    fx = f.subs(x, x0)  # f(x0) [Evaluating]
    fp = f.diff(x)  # f'(x)
    fpx = fp.subs(x, x0) # f'(x0) [Evaluating]
    fpp = fp.diff(x)  # f''(x)
    fppx = fpp.subs(x, x0)  # f''(x0) [Evaluating]
    cont = 1
    # print(f)
    # print(fx)
    # print(fp)
    # print (fpx)
    # print(fpp)
    # print(fppx)
    denominator = (fpx ** 2) - (fx * fppx)
    err = tolerance + 1
    inicial=x0
    nCurrent = 0
    print(". . .")
    # ITERATING
    while denominator != 0 and fx != 0 and err > tolerance and cont < nIterations:
        print("i:{:03d} x: {:.10f} fx: {:.10f} error abs: {:.10f} \n"
              .format(cont,float(x0) , float(fx), float(err)), end="")
        x1 = x0 - (fx * fpx) / denominator
        fx = f.subs(x, x1)
        fpx = fp.subs(x, x1)
        fppx = fpp.subs(x, x1)
        denominator = (fpx ** 2) - (fx * fppx)
        err = abs(x1 - x0)
        x0 = x1
        cont = cont + 1
    print("i:{:03d} x: {:.10f} fx: {:.10f} error abs: {:.10f} \n"
          .format(cont, float(x0), float(fx), float(err)), end="")
    # RESULT
    if (fx == 0):
        print(float(x0), "Es una raíz")
    else:
        if (err < tolerance):
            print(float(x1), "Es aproximación a una raíz con tolerancia de", tolerance)
        else:
            if (denominator == 0):
                print("Se encontró una división por 0")
            else:
                print("El método fracasó brutalmente con", nIterations, "iteraciones")

    sp.plot(f, (x, inicial, x0 + nIterations * tolerance))